File: linked-list/doubly linklist/doubleLinklist.go
package main

import "fmt"

// DLNode represents a node in the doubly linked list.
type DLNode struct {
	Val  int
	Prev *DLNode
	Next *DLNode
}

// Solution is a struct that manages the doubly linked list.
type Solution struct {
	head *DLNode
}

// Insert inserts a new node with the given Val at the front of the list.
func (s *Solution) Insert(Val int) {
	newNode := &DLNode{Val: Val}
	newNode.Next = s.head
	if s.head != nil {
		s.head.Prev = newNode
	}
	s.head = newNode
}

// InsertAfter inserts a new node with the given Val after the provided prevNode.
func (s *Solution) InsertAfter(prevNode *DLNode, Val int) {
	if prevNode == nil {
		fmt.Println("The given previous node cannot be nil") // Check if the previous node exists
		return
	}
	newNode := &DLNode{Val: Val}
	newNode.Next = prevNode.Next
	prevNode.Next = newNode
	newNode.Prev = prevNode
	if newNode.Next != nil {
		newNode.Next.Prev = newNode
	}
}

// Delete removes the node with the given Val from the list.
func (s *Solution) Delete(key int) {
	temp := s.head
	for temp != nil {
		if temp.Val == key {
			if temp.Prev != nil {
				temp.Prev.Next = temp.Next
			} else {
				s.head = temp.Next
			}
			if temp.Next != nil {
				temp.Next.Prev = temp.Prev
			}
			return
		}
		temp = temp.Next
	}
}

// Search searches for a node with the given Val and returns true if found, false otherwise.
func (s *Solution) Search(key int) bool {
	current := s.head
	for current != nil {
		if current.Val == key {
			return true // Return true if the key is found in the list
		}
		current = current.Next
	}
	return false // Return false if the key is not found in the list
}

func main() {
	dll := &Solution{}
	dll.Insert(1)
	dll.Insert(2)
	dll.Insert(3)

	fmt.Println("Search 2:", dll.Search(2)) // true
	dll.Delete(2)
	fmt.Println("Search 2:", dll.Search(2)) // false
}

File: linked-list/doubly linklist/doubleLinklist.java
class DLNode {
    int val;
    DLNode prev;
    DLNode next;

    // Constructor to initialize a DLNode with a given value.
    DLNode(int d) {
        val = d;
        prev = next = null;  // Initialize previous and next pointers to null.
    }
}

public class doubleLinklist {
  DLNode head;

    // Inserts a new DLNode at the front of the list.
    void insert(int val) {
        DLNode newNode = new DLNode(val);  // Create a new DLNode with the given value.
        newNode.next = head;
        if (head != null)
            head.prev = newNode;
        head = newNode;  // Update the head to the new DLNode.
    }

    // Inserts a new node after the given prev_node.
    void insertAfter(DLNode prev_node, int val) {
        if (prev_node == null) {
            System.out.println("The given previous node cannot be null");  // Check if the previous node is null.
            return;
        }
        DLNode newNode = new DLNode(val);  // Create a new DLNode with the given value.
        newNode.next = prev_node.next;
        prev_node.next = newNode;
        newNode.prev = prev_node;
        if (newNode.next != null)
            newNode.next.prev = newNode;
    }

    // Deletes the node with the given val.
    void delete(int key) {
        DLNode temp = head;
        while (temp != null) {
            if (temp.val == key) {
                if (temp.prev != null)
                    temp.prev.next = temp.next;  // Adjust pointers to remove the DLNode.
                else
                    head = temp.next;
                if (temp.next != null)
                    temp.next.prev = temp.prev;
                return;
            }
            temp = temp.next;
        }
    }

    // Searches for the node with the given val.
    boolean search(int key) {
        DLNode current = head;
        while (current != null) {
            if (current.val == key)
                return true;  // Return true if the key is found in the list.
            current = current.next;
        }
        return false;  // Return false if the key is not found in the list.
    }

    public static void main(String[] args) {
        doubleLinklist dll = new doubleLinklist();
        dll.insert(1);
        dll.insert(2);
        dll.insert(3);

        System.out.println("Search 2: " + dll.search(2));  // true
        dll.delete(2);
        System.out.println("Search 2: " + dll.search(2));  // false
    }
}
